File: java/src/main/java/com/purescraps/purehtml/transformers/types/ReplaceTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.util.List;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

/**
 * replace: Replace regex/string matches in a string
 *
 * Usage:
 * replace(pattern, replacement): replaces every match of `pattern` (a
 * regular expression) with `replacement`. `replacement` may reference
 * capture groups using $1, $2, etc.
 */
public class ReplaceTransformer extends Transformer {
    private List<String> args;

    public ReplaceTransformer(List<String> args) {
        this.args = args;
    }

    public ReplaceTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    public String getName() {
        return "replace";
    }

    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (!(val instanceof String)) {
            throw new IllegalArgumentException("replace: expected string. Got " + val);
        }

        if (args == null || args.size() < 2) {
            throw new IllegalArgumentException(
                    "replace: requires a pattern and a replacement, e.g. replace(\\s+, ' ')");
        }

        try {
            Pattern pattern = Pattern.compile(args.get(0));
            return pattern.matcher((String) val).replaceAll(args.get(1));
        } catch (PatternSyntaxException e) {
            throw new IllegalArgumentException("replace: invalid pattern: " + args.get(0), e);
        }
    }
}

File: python/purehtml/transformers/types/replace_transformer.py
import re
from abc import ABC
from typing import List

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class ReplaceTransformer(Transformer, ABC):
    """
    replace: Replace regex/string matches in a string

    Usage:
    replace(pattern, replacement): replaces every match of `pattern` (a
    regular expression) with `replacement`. `replacement` may reference
    capture groups using \\1, \\2, etc.
    """

    def __init__(self, args: List[str] = None):
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        self.args = args

    def get_name(self) -> str:
        return "replace"

    def transform(self, params: TransformParams):
        val = params.get_val()

        if val is None:
            return None

        if not isinstance(val, str):
            raise TypeError(f"replace: expected string. Got {val!r}")

        if len(self.args) < 2:
            raise ValueError(
                "replace: requires a pattern and a replacement, e.g. replace(\\s+, ' ')"
            )

        pattern, replacement = self.args[0], self.args[1]

        try:
            return re.sub(pattern, replacement, val)
        except re.error as e:
            raise ValueError(f"replace: invalid pattern: {pattern!r}") from e
